go/plugins/googleai/googleai.go

```diff
@@ -16,17 +16,154 @@ package googleai
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("googleai.Init: %w", err)
+		}
+	}()
+
+	if err := initClient(ctx, cfg.APIKey); err != nil {
+		return err
+	}
+
+	needModels := len(cfg.Models) == 0
+	needEmbedders := len(cfg.Embedders) == 0
+	if needModels || needEmbedders {
+		iter := state.client.ListModels(ctx)
+		for {
+			mi, err := iter.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return err
+			}
+			// Model names are of the form "model/name".
+			name := path.Base(mi.Name)
+			if needModels && slices.Contains(mi.SupportedGenerationMethods, "generateContent") {
+				cfg.Models = append(cfg.Models, name)
+			}
+			if needEmbedders && slices.Contains(mi.SupportedGenerationMethods, "embedContent") {
+				cfg.Embedders = append(cfg.Embedders, name)
+			}
+		}
+	}
+	state.models = map[string]ai.Generator{}
+	for _, name := range cfg.Models {
+		state.models[name] = defineModel(name, state.client)
+	}
+
+	state.embedders = map[string]ai.Embedder{}
+	for _, name := range cfg.Embedders {
+		state.embedders[name] = defineEmbedder(name, state.client)
+	}
+	return nil
+}
+
+func initClient(ctx context.Context, apiKey string) error {
+	if apiKey == "" {
+		return errors.New("missing API key")
+	}
+
+	state.mu.Lock()
+	defer state.mu.Unlock()
+	if state.client != nil {
+		return errors.New("already initialized")
+	}
+	c, err := genai.NewClient(ctx, option.WithAPIKey(apiKey))
+	if err != nil {
+		return err
+	}
+	state.client = c
+	return nil
+}
+
+func defineModel(name string, client *genai.Client) ai.Generator {
+	meta := &ai.GeneratorMetadata{
+		Label: "Google AI - " + name,
+		Supports: ai.GeneratorCapabilities{
+			Multiturn: true,
+		},
+	}
+	g := generator{model: name, client: client}
+	return ai.DefineGenerator("google-genai", name, meta, g.Generate)
+}
+
+func defineEmbedder(name string, client *genai.Client) ai.Embedder {
+	return ai.DefineEmbedder("google-genai", name, func(ctx context.Context, input *ai.EmbedRequest) ([]float32, error) {
+		em := client.EmbeddingModel(name)
+		parts, err := convertParts(input.Document.Content)
+		if err != nil {
+			return nil, err
+		}
+		res, err := em.EmbedContent(ctx, parts...)
+		if err != nil {
+			return nil, err
+		}
+		return res.Embedding.Values, nil
+	})
+}
+
+// Generator returns the generator with the given name.
+// It panics if the generator was not configured.
+func Generator(name string) ai.Generator {
+	if g, ok := state.models[name]; ok {
+		return g
+	}
+	panic(fmt.Sprintf("googleai: generator %q not configured", name))
+}
+
+// Embedder returns the embedder with the given name.
+// It panics if the embedder was not configured.
+func Embedder(name string) ai.Embedder {
+	if e, ok := state.embedders[name]; ok {
+		return e
+	}
+	panic(fmt.Sprintf("googleai: embedder %q not configured", name))
+}
+
+// Generators returns the names of the configured generators.
+func Generators() []string {
+	return maps.Keys(state.models)
```

Comment:
At this point all the generators are actually defined.  So why not return them, rather than returning strings and requiring the user to call `Generator`?

Or, another approach would be to have `Generator` call `defineModel` itself if necessary, using the mutex and client in `state`.

Reply:
> At this point all the generators are actually defined. So why not return them, rather than returning strings and requiring the user to call Generator?

Maybe. We'd also need a way to get the name from a Generator, in case someone wanted it.

> Or, another approach would be to have Generator call defineModel itself if necessary, using the mutex and client in state.

That brings us back to the problem that Init is trying to solve, which is to register all actions up front so the genkit UI can show them. Your way, like my original idea, would make that list control-flow dependent.

Reply:
Or maybe `Generators` returns (a copy of) the internal map in state. In which case who needs `Generate`?

The reason I added `Generators` in the first place was to handle the case where you provide the empty list of models and let the List RPC fill it in for you. There would be no other way to know what models were available. That's also why I just returned names; you only care about the names for that use case.

Reply:
I removed `Generators` and `Embedders` for now.